File: apps/server/src/lib/file-system/file-system.ts
import { FileSystemEnum } from 'database';
import { LocalFileSystem } from './local-file-system.js';
import { MinIOFileSystem } from './minio-file-system.js';

export interface IFileSystem {
  createDir(dirPath: string): Promise<void>;
  exists(filePath: string): Promise<boolean>;
  read(filePath: string): Promise<Buffer>;
  write(filePath: string, content: string): Promise<void>;
}

class FileSystem {
  private fileSystems: { [key in FileSystemEnum]: IFileSystem } = {
    [FileSystemEnum.LOCAL]: new LocalFileSystem(),
    [FileSystemEnum.MINIO]: new MinIOFileSystem(),
  };

  async createDir(fileSystem: FileSystemEnum, path: string): Promise<void> {
    this.fileSystems[fileSystem].createDir(path);
  }

  async exists(fileSystem: FileSystemEnum, filePath: string): Promise<boolean> {
    return this.fileSystems[fileSystem].exists(filePath);
  }

  async read(fileSystem: FileSystemEnum, filePath: string): Promise<Buffer> {
    return this.fileSystems[fileSystem].read(filePath);
  }

  async write(fileSystem: FileSystemEnum, filePath: string, content: string): Promise<void> {
    this.fileSystems[fileSystem].write(filePath, content);
  }
}

export const fileSystem = new FileSystem();

File: apps/ingest/common/file_system/FileSystem.py
import sqlmodel

from common.file_system.LocalFileSystem import LocalFileSystem
from common.file_system import FileSystemEnum
from common.db import engine
from common.file_system.MinIOFileSystem import MinIOFileSystem

from model.db.File import File


class FileSystem:
    """
    Wraps the different file systems we have (e.g LocalFileSystem)
    You pass in a File object and based on the FileSystem it uses we delegate the read/write operations
    """
    def __init__(self):
        self.file_systems = {
            FileSystemEnum.LOCAL: LocalFileSystem(),
            FileSystemEnum.MINIO: MinIOFileSystem()
        }

    def create_dir(self, file_system: FileSystemEnum, path):
        return self.file_systems[file_system].create_dir(path)

    def exists(self, file: File):
        return self.file_systems[file.file_system].exists(file)

    def read(self, file: File):
        return self.file_systems[file.file_system].read(file)

    def write(self, file_system: FileSystemEnum, external_id: str, content) -> File:
        file = File(external_id=external_id, file_system=file_system)

        # write to file system
        self.file_systems[file.file_system].write(file, content)

        # save to database
        session = sqlmodel.Session(engine)

        with session:
            session.add(file)
            session.commit()
            session.refresh(file)

        return file
